File: FE/lib/services/location_service.dart
import 'api_service.dart';

class LocationOption {
  const LocationOption({required this.code, required this.name});

  final int code;
  final String name;
}

class LocationService {
  LocationService(this._apiService);

  final ApiService _apiService;

  Future<List<LocationOption>> fetchProvinces() async {
    final response = await _apiService.get('/locations/provinces');
    final data = (response as Map<String, dynamic>)['data'] as Map<String, dynamic>?;
    final list = (data?['provinces'] as List<dynamic>? ?? const <dynamic>[]);

    return list
        .map(
          (item) => LocationOption(
            code: (item['code'] as num).toInt(),
            name: item['name'] as String,
          ),
        )
        .toList(growable: false);
  }

  Future<List<LocationOption>> fetchDistricts(int provinceCode) async {
    final response = await _apiService.get('/locations/provinces/$provinceCode/districts');
    final data = (response as Map<String, dynamic>)['data'] as Map<String, dynamic>?;
    final list = (data?['districts'] as List<dynamic>? ?? const <dynamic>[]);

    return list
        .map(
          (item) => LocationOption(
            code: (item['code'] as num).toInt(),
            name: item['name'] as String,
          ),
        )
        .toList(growable: false);
  }

  Future<List<LocationOption>> fetchWards(int districtCode) async {
    final response = await _apiService.get('/locations/districts/$districtCode/wards');
    final data = (response as Map<String, dynamic>)['data'] as Map<String, dynamic>?;
    final list = (data?['wards'] as List<dynamic>? ?? const <dynamic>[]);

    return list
        .map(
          (item) => LocationOption(
            code: (item['code'] as num).toInt(),
            name: item['name'] as String,
          ),
        )
        .toList(growable: false);
  }
}

File: BE/src/services/location.service.js
import env from '../config/env.js';
import AppError from '../utils/appError.js';

const requestJson = async (path) => {
  const response = await fetch(`${env.locationApiBase}${path}`);

  if (!response.ok) {
    throw new AppError('Failed to fetch location data', 502);
  }

  return response.json();
};

export const getProvinces = async () => {
  const provinces = await requestJson('/p/');
  return provinces.map((item) => ({ code: item.code, name: item.name }));
};

export const getDistrictsByProvince = async (provinceCode) => {
  const province = await requestJson(`/p/${provinceCode}?depth=2`);
  return (province.districts || []).map((item) => ({ code: item.code, name: item.name }));
};

export const getWardsByDistrict = async (districtCode) => {
  const district = await requestJson(`/d/${districtCode}?depth=2`);
  return (district.wards || []).map((item) => ({ code: item.code, name: item.name }));
};
